### ISSUE TITLE
브랜치 전략: feature-infer-ens-submit

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#12]
## 목표와 범위
- **목표:**
    검증된 단일/다중 모델을 대상으로 **안정적·재현 가능한 추론 파이프라인**을 구축하고, **TTA/캘리브레이션/가중 앙상블**을 통해 Macro-F1을 극대화하며, **제출 파일(pred.csv) 자동 검증**을 표준화합니다.
    
    - 본 브랜치는 **추론 전처리·TTA 정책·로짓/확률 집계·모델 가중 탐색·캘리브레이션(Validation 기반)·제출물 생성/검수·로그/리포트**를 담당합니다.
    - 모든 기능은 **`ipynb`에서 사전 검증** 후, 통과 코드만 `src/`의 표준 디렉터리로 **모듈화 이관**합니다.
    - 병합 전 **정확성(형식·ID·라벨 도메인), 성능(검증 점수), 재현성(고정된 결과), 운영성(속도/자원) 기준**을 완료 조건(DoD)으로 정의합니다.
- **범위:**
    - **싱글/멀티 모델 추론**(GPU/CPU), **TTA**(회전/flip/경미 색상), **집계**(logit 평균 권장), **캘리브레이션**(온도 스케일링; 검증셋 전용), **가중 앙상블**(simplex grid/좌표상승/순위평균 대안), **제출 검증**(ID/행수/범위/중복/정렬).
    - **규칙 준수:** 테스트 데이터로 **학습/튜닝 금지**, 외부 데이터/유료 API 금지, Pseudo Labeling 금지.

---

### 1) 추론 구성·레지스트리
- **모델 레지스트리(`ensemble_config.json`)**
    - 각 항목: `name, arch, checkpoint, img_size, mean/std, tta_profile, weight(init), temperature(optional)`
    - 체크포인트 경로는 `experiments/train/.../best.pth` 또는 `experiments/robust/.../best.pth`
- **추론 설정(`infer_config.json`)**
    - 공통 전처리(Val/Test와 동일), 배치/워커/디바이스, TTA 예산, logit/확률 집계방식, 출력 경로, 엄격 모드(strict), 시드

### 2) TTA 정책(테스트 전용)
- **기본 세트:** 회전 {0°,90°,180°,270°} × HFlip {False,True(선택)} = 4~8 views
- **경미 색상(옵션 1종):** 밝기 ±5% 중 1 view만 추가 가능(과도 금지)
- **집계:** **logit 평균**(권장) → 소프트맥스; 확률 평균/기하 평균 옵션 제공
- **오리엔테이션 선택기(옵션):** 회전별 최대 신뢰도 비교 후 최빈 회전 채택(다수결)

### 3) 캘리브레이션(검증셋 전용)
- **Temperature Scaling(권장):**
    - 각 모델별 온도 T 최적화(검증 Cross-Entropy 최소화)
    - 저장: `temperature`를 `ensemble_config.json`에 반영
- **Isotonic(비권장):** 데이터 적음/과적합 위험으로 기본 미사용
- **주의:** 테스트 데이터로 **온도 학습 금지**

### 4) 앙상블 전략
- **가중 평균(확장):**
    - 제약: `w_i ≥ 0`, `∑ w_i = 1`
    - 탐색: **simplex grid**(step=0.05 또는 0.02), **좌표상승(coordinate ascent)**, **베스트 K 모델 subset** 탐색
    - 목적함수: **검증 Macro-F1**(OOF 또는 검증셋) 최대화
- **대안:**
    - **Rank Averaging**: 각 모델의 클래스 순위를 평균 → 스케일 불일치 견고
    - **Soft-Vote vs Logit-Vote**: logit 기반이 보통 더 안정적(권장)

### 5) 제출 전·후 검증
- **전:**
    - 입력 `sample_submission.csv`와 테스트 이미지 디렉토리 존재/행수 일치 확인
    - ID 정렬 순서 **원본 유지**(추론 시 임의 셔플 금지)
    - 확률 벡터 **NaN/Inf 없음**, sum≈1(±1e-6) 재정규화
- **후:**
    - `pred.csv` 행수=3,140, 컬럼=`['ID','target']`
    - `ID` 집합 = `sample_submission.csv`와 **완전 동일**
    - `target ∈ {0..16}`, 결측/중복 없음
    - 해시/체크섬 로그 남김

### 6) 성능·운영
- **배치·워커 튜닝:** `batch_size, num_workers, pin_memory, persistent_workers`
- **정밀도:** 추론 FP16 가능(메모리 절약) + **집계/캘리브레이션은 FP32**로 수행(수치 안정성)
- **스루풋 로그:** imgs/sec 측정(워밍업 제외), TTA off/on 모두 기록
- **다중 GPU(옵션):** `torch.nn.DataParallel` 또는 간단 샤딩(선택)

### 7) 로깅·리포팅
- **실험 로그:** `experiments/infer/<run_id>/{config.json, ensemble_config.json, timings.json, calibration.json, weight_search.csv, val_scores.csv}`
- **그림/리포트:** `weight_vs_score_curve.png`, `tta_vs_score_curve.png`, `docs/infer_report_<timestamp>.html`
- **제출 산출:** `submissions/pred_<tag>_<timestamp>.csv` (+ `pred.sha256`)

---

## ipynb-우선 워크플로우 (파일명·셀 단위 체크리스트)
1. **`notebooks/INFER/00_infer_config_and_registry.ipynb`**
    - [ ]  `configs/infer_config.json`, `configs/ensemble_config.json` 작성/검증
    - [ ]  체크포인트 존재/로드 스모크 테스트
2. **`notebooks/INFER/01_single_model_tta_sweep.ipynb`**
    - [ ]  모델 1~2개 대상 TTA(4/6/8 views) 스윕 → 검증 Macro-F1 vs 시간 곡선
    - [ ]  logit 평균 vs 확률 평균 비교
3. **`notebooks/INFER/02_calibration_temperature_scaling.ipynb`**
    - [ ]  각 모델별 Temperature T 학습(검증 셋) → `calibration.json` 저장
    - [ ]  Calibration 전/후 ECE/Macro-F1 비교(검증)
4. **`notebooks/INFER/03_weight_search_and_rank_averaging.ipynb`**
    - [ ]  simplex grid/좌표상승으로 가중 벡터 탐색 → `weight_search.csv`
    - [ ]  Rank Averaging 대안 결과 비교
5. **`notebooks/INFER/04_ensemble_eval_and_selection.ipynb`**
    - [ ]  최종 후보 3~5개 조합 검증 → 베스트 1안 선택(성과/시간/안정성)
    - [ ]  리포트용 표/그림 생성
6. **`notebooks/INFER/05_make_submission_and_validation.ipynb`**
    - [ ]  테스트 세트 추론(TTA/캘리브레이션/가중 적용) → `pred.csv` 생성
    - [ ]  형식 검증(행수/ID/범위/중복/해시) 및 로그 저장
7. **`notebooks/INFER/06_export_inference_modules.ipynb`**
    - [ ]  통과 코드만 `src/inference/*.py`, `src/utils/*.py`로 이관
    - [ ]  `__all__`/docstring/타입힌트/CLI 사용 예시 정리

---

## 산출물(Artifacts)
- **코드 모듈**
    - `src/inference/loader.py` : 테스트 로더(정렬/반복불가), 전처리
    - `src/inference/tta.py` : TTA 뷰 생성/복원, 집계(logit/prob/geomean)
    - `src/inference/predict.py` : 단일 모델 추론(캘리브레이션 포함)
    - `src/inference/ensemble.py` : 가중 평균/좌표상승/순위평균, weight search
    - `src/inference/make_submission.py` : 엔드투엔드 실행(검증→추론→검수→저장)
    - `src/inference/validate_submit.py` : `pred.csv` 형식/정합성 검사
    - `src/inference/registry.py` : 모델/체크포인트/온도/tta 프로필 레지스트리
    - (연계) `src/utils/metrics.py` : Macro-F1, ECE(추가)
- **리포트/로그**
    - `docs/infer_report_<timestamp>.html`
    - `experiments/infer/<run_id>/{config.json, ensemble_config.json, calibration.json, timings.json, weight_search.csv, val_scores.csv, tta_vs_score.png, weight_vs_score.png}`
- **설정 파일**
    - `configs/infer_config.json`, `configs/ensemble_config.json`
- **노트북(최종 파일명 확정)**
    - `notebooks/INFER/00_infer_config_and_registry.ipynb`
    - `notebooks/INFER/01_single_model_tta_sweep.ipynb`
    - `notebooks/INFER/02_calibration_temperature_scaling.ipynb`
    - `notebooks/INFER/03_weight_search_and_rank_averaging.ipynb`
    - `notebooks/INFER/04_ensemble_eval_and_selection.ipynb`
    - `notebooks/INFER/05_make_submission_and_validation.ipynb`
    - `notebooks/INFER/06_export_inference_modules.ipynb`

---

### 1 추론 파이프라인

- 입력: `sample_submission.csv`, `data/raw/test`, `ensemble_config.json`, `infer_config.json`
- 출력: `submissions/pred_<tag>_<timestamp>.csv` (+ 해시)
- 요구: **ID 정렬 보존**, 배치/워커 튜닝, FP16 추론 + FP32 집계, 드롭아웃/BatchNorm eval 모드

### 2 TTA·집계
- 최소 4 views(회전), 최대 8 views(+HFlip) 권장
- 집계는 **logit 평균** 기본, 확률 평균/기하 평균 옵션
- TTA 순서/시드 고정 → 재실행 동일 결과

### 3 캘리브레이션
- Temperature Scaling per model(검증셋), T 저장/재사용
- 적용 순서: **(모델 로짓 → /T → 집계)** 또는 **(모델 확률 → 집계)** 중 하나로 일관성 유지

### 4 앙상블
- 가중 탐색 제약/탐색자 제공, 베스트 가중 및 대안(rank) 저장
- 모델 누락/로드 실패 시 **즉시 실패**(strict=true), lenient=false 기본

### 5 제출 검증
- `len(pred) == len(sample_submission)`
- `set(ID)` 동일, `target ∈ [0..16]` 정수, 중복/결측 없음
- CSV 인코딩/개행 규격 표준(CRLF/LF 허용, BOM 미삽입 권장)

### 6 성능·운영
- 스루풋 측정(워밍업 제외), TTA off/on 로깅
- 메모리 피크 모니터링(옵션), OOM 시 배치 자동 축소

---

## 실험·검증(Validation) 기준
| 항목 | 측정/검증 방법 | 통과 기준(제안) |
| --- | --- | --- |
| **형식 정확성** | `validate_submit.py` | ID/행수/범위/중복/정렬 **모두 통과** |
| **성능 향상(TTA)** | 01 노트북 | TTA on이 off 대비 **+0.3~1.0%p**(검증 Macro-F1) |
| **성능 향상(앙상블)** | 04 노트북 | 앙상블이 최고 단일 모델 대비 **≥ +0.5%p** |
| **캘리브레이션 효과** | 02 노트북 | ECE 감소, Macro-F1 **동일↑** |
| **재현성** | 동일 설정 2회 | 검증 점수 편차 **≤ ±0.3%p** |
| **스루풋** | 00/05 노트북 | 팀 합의 기준 **≥ X imgs/s** (TTA off 기준) |
| **안전성** | 로깅 검사 | NaN/Inf 0건, 예외 0건 |
> X imgs/s는 장비 측정 후 고정합니다(프로젝트 표준과 동일).
> 

---

## 병합 조건(Definition of Done & PR 체크리스트)
- **DoD**
    - ✅ 00~06 노트북 **모든 셀 실행 완료(오류 0, 출력 포함)**
    - ✅ `docs/infer_report_*.html` 1부 생성(가중/캘리브레이션/TTA 결과 포함)
    - ✅ `pred.csv` 생성 및 `validate_submit.py` **통과**
    - ✅ 검증 기준 충족(표 상기), 재현성 확인
    - ✅ 규칙 준수(테스트 데이터로 학습/튜닝 금지, 외부 데이터/유료 API 미사용)
- **PR 체크리스트**
    - [ ]  코드 포맷팅(black/isort), 린트 통과
    - [ ]  단위 테스트 4종 통과(캘리브레이션·TTA 집계·가중 탐색·제출 검증)
    - [ ]  `configs/infer_config.json`, `configs/ensemble_config.json` 커밋
    - [ ]  아티팩트 링크(로그/리포트/제출 파일 해시) 첨부

---

## 브랜치 작업 순서(권장 커밋 플로우)
1. `feat(config)`: infer/ensemble config 스켈레톤 + 로더 스모크
2. `feat(tta)`: TTA 프로필/집계 구현 + 01 노트북
3. `feat(calib)`: Temperature Scaling + 02 노트북
4. `feat(ens)`: 가중 탐색/순위평균 + 03·04 노트북
5. `feat(submit)`: make_submission/validate_submit + 05 노트북
6. `feat(report)`: infer_report HTML 생성
7. `feat(export)`: 모듈 이관 + 타입힌트/도큐
8. `tests/*`: 유닛테스트 추가

---

## 구성·설정 표준
**`configs/infer_config.json` (예시)**

```json
{
  "img_size": 224,
  "batch_size": 128,
  "num_workers": 4,
  "pin_memory": true,
  "persistent_workers": true,
  "use_fp16": true,
  "aggregate": "logit_mean",
  "tta": { "rotations": [0,90,180,270], "hflip": true, "light_color_jitter": false },
  "paths": {
    "test_root": "data/raw/test",
    "sample_submission": "data/raw/sample_submission.csv",
    "exp_dir": "experiments/infer",
    "submit_dir": "submissions"
  },
  "strict": true,
  "seed": 42,
  "tag": "core_tta8"
}
```

**`configs/ensemble_config.json` (예시)**

```json
{
  "models": [
    {
      "name": "resnet50_224",
      "arch": "resnet50",
      "checkpoint": "experiments/train/20250901_resnet50_img224_bs64/best.pth",
      "img_size": 224,
      "mean": [0.485, 0.456, 0.406],
      "std":  [0.229, 0.224, 0.225],
      "tta_profile": "rot4_hflip",
      "weight": 0.40,
      "temperature": 1.20
    },
    {
      "name": "effb3_300",
      "arch": "efficientnet_b3",
      "checkpoint": "experiments/train/20250902_effb3_img300_bs48/best.pth",
      "img_size": 300,
      "mean": [0.485, 0.456, 0.406],
      "std":  [0.229, 0.224, 0.225],
      "tta_profile": "rot4",
      "weight": 0.35,
      "temperature": 1.05
    },
    {
      "name": "convnext_t_288",
      "arch": "convnext_tiny",
      "checkpoint": "experiments/robust/20250903_convnext_t_robust/best.pth",
      "img_size": 288,
      "mean": [0.485, 0.456, 0.406],
      "std":  [0.229, 0.224, 0.225],
      "tta_profile": "rot4_hflip",
      "weight": 0.25,
      "temperature": 1.10
    }
  ]
}
```

---

## 파라미터 표준(요약)
| 항목 | 기본값 | 허용 범위 | 비고 |
| --- | --- | --- | --- |
| TTA 회전 | 0/90/180/270 | 0~4개 | 최소 rot4 권장 |
| TTA HFlip | false | true/false | 문서 도메인 특성 상 과다 금지 |
| 집계 방식 | logit_mean | prob_mean/geomean | logit 권장 |
| Temperature T | 1.0 | 0.8~2.0 | 검증셋 최적화 |
| 가중 탐색 step | 0.05 | 0.02~0.1 | simplex grid |
| 배치 크기 | 128 | 32~256 | 메모리 기반 조정 |
| 정밀도 | fp16+fp32 agg | fp32 | agg는 FP32 고정 |

---

## 실패 정책(Failure Policy)
- **형식 오류**(행수/ID/범위 불일치) → 즉시 실패, `errors.log` 기록, `pred.csv` 폐기
- **로드 실패**(모델/체크포인트 손상) → strict 모드에서 즉시 실패; lenient=false 기본
- **수치 이상**(NaN/Inf) → 해당 배치 재시도 후 지속 시 중단
- **성능 미달**(앙상블 < 단일 최고) → 가중 재탐색 또는 단일 모델 제출 플랜 자동 제시

---

## 리스크·준수사항
- **규칙 준수:** 테스트 데이터 **학습/튜닝**에 사용 금지(캘리브레이션·가중 탐색은 검증셋만), 외부 데이터/유료 API 미사용, Pseudo Labeling 금지.
- **운영 리스크:** TTA/다중 해상도 혼합 시 속도·메모리 급증 → 배치/해상도/뷰수 예산화.
- **스케일 리스크:** 서로 다른 해상도/정규화 혼합 → 각 모델별 전처리 파이프라인 철저 분리 후 집계.
- **안정성 리스크:** FP16 underflow → **집계/캘리브레이션은 FP32**로 강제.
